### PR TITLE
Kt task/user overview

### PR DESCRIPTION
**Pull request by borkarsaish65**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: controllers/v1/userProjects.js [hunk 1]
```diff
@@ -1125,5 +1126,83 @@ module.exports = class UserProjects extends Abstract {
...
+    /**
+
+    /**
+     * Lists projects based on the status
+     * @method
+     * @name listPendingProjects
+     * @returns {JSON} Project details
+    */
+
+    async listPendingProjects(req) {
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1145]: we can have only one controller @borkarsaish65

File: controllers/v1/userProjects.js [hunk 2]
```diff
@@ -1125,5 +1126,83 @@ module.exports = class UserProjects extends Abstract {
...
+    /**
+     * Lists projects created by the userId
+     * @method
+     * @name listCreatedProjects
+     * @returns {JSON} Project details
+    */
+
+
+    async listCreatedProjects(req) {
+        return new Promise(async (resolve, reject) => {
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1186]: wen dont need this controller @borkarsaish65

File: controllers/v1/userProjects.js (hunk 1 above)
Comment (ankit1shahu) [reply controllers/v1/userProjects.js:1145]: controller can named as usersProject

File: generics/services/users.js [hunk 3]
```diff
@@ -89,6 +89,7 @@ const locationSearch = function ( filterData, formatResult = false ) {
         bodyData["request"]["filters"] = filterData;
         const url = 
         userServiceUrl + CONSTANTS.endpoints.GET_LOCATION_DATA;
+        console.log(url,'url')
```
Comment (ankit1shahu) [inline generics/services/users.js:92]: remove console logs @borkarsaish65

File: module/project/helper.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/**
+ * name : helper.js
+ * author : Saish
+ * created-date : 19-june-2024
+ * Description : Projects helper functionality.
+ */
+
```
Comment (ankit1shahu) [inline module/project/helper.js:7]: why this file is created @borkarsaish65

**Review by ankit1shahu — CHANGES_REQUESTED** [review]

File: controllers/v1/userProjects.js (hunk 1 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js:1145]: I could do that but I used the same code which I had created for previous task given by you. So there we were asked list projects based on the type passed so its the same.

The second point of creating two different controller is that one is queried by userId field and another is queried by createdBy field.

Let me know what you think ?

File: controllers/v1/userProjects.js (hunk 2 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js:1186]: explaination given above

File: generics/services/users.js (hunk 3 above)
Comment (borkarsaish65) [reply generics/services/users.js:92]: Done

File: module/project/helper.js (hunk 4 above)
Comment (borkarsaish65) [reply module/project/helper.js:7]: Removed

File: controllers/v1/userProjects.js (hunk 1 above)
Comment (ankit1shahu) [reply controllers/v1/userProjects.js:1145]: no that not its should work and now we have only two task 
big number and project details that can be done in one API 
create only one controller @borkarsaish65

Comment (borkarsaish65) [reply controllers/v1/userProjects.js:1145]: Done

File: controllers/v1/userProjects.js [hunk 5]
```diff
@@ -1125,5 +1126,106 @@ module.exports = class UserProjects extends Abstract {
...
+    /**
+     * Lists projects based on the status
+     * @method
+     * @name listPendingProjects
+     * @returns {JSON} Project details
+    */
+
+    async listPendingProjects(req) {
+        return new Promise(async (resolve, reject) => {
+
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1147]: merge this into one controller as userOverview don't list pending or started or inprogress @borkarsaish65

**Review by ankit1shahu — CHANGES_REQUESTED** [review]

File: controllers/v1/userProjects.js (hunk 5 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js]: done

File: controllers/v1/userProjects.js [hunk 6]
```diff
@@ -1125,5 +1126,29 @@ module.exports = class UserProjects extends Abstract {
                 }
         })
     }
+    async listUserProjects(req) {
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1129]: rename this controller to userProjects @borkarsaish65

File: module/userProjects/validator/v1.js [hunk 7]
```diff
@@ -33,6 +33,19 @@ module.exports = (req) => {
             req.checkBody("data").exists().withMessage("data is required");
             req.checkBody("data.transactionId").exists().withMessage("transactionId is required");
             req.checkBody("data.osid").exists().withMessage("osid is required");
+        },
+        listUserProjects:function(){
+            req.checkQuery('userInvolvement')
```
Comment (ankit1shahu) [inline module/userProjects/validator/v1.js:38]: what is this userInvolvment? @borkarsaish65

File: module/userProjects/helper.js [hunk 8]
```diff
@@ -3273,6 +3273,75 @@ module.exports = class UserProjectsHelper {
...
+            }
+          });
+    }
+    static listUserProjects({userId,userInvolvement,stats}){
+
+        return new Promise(async (resolve, reject) => {
+            try{
+                switch(userInvolvement)
+                {
+                    case 'creator':
```
Comment (ankit1shahu) [inline module/userProjects/helper.js:3320]: the projects collection will have projects consumed by user there is no difference that project creator or consumed both are same only @borkarsaish65

**Review by ankit1shahu — CHANGES_REQUESTED** [review]

File: controllers/v1/userProjects.js (hunk 6 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js]: done

File: module/userProjects/validator/v1.js (hunk 7 above)
Comment (borkarsaish65) [reply module/userProjects/validator/v1.js]: I have removed the logic now only projects which are consumed will appear

File: module/userProjects/helper.js (hunk 8 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js]: removed the logic

File: controllers/v1/userProjects.js [hunk 9]
```diff
@@ -1125,5 +1126,33 @@ module.exports = class UserProjects extends Abstract {
...
+            try {
+
+                if (req.query.stats) {
+                    //convert req.query.stats string to Boolean value
+                    req.query.stats = UTILS.convertStringToBoolean(req.query.stats);
+                }
+
+                let listOfCreatedProjects = await userProjectsHelper.listUserProjects({
+                    userId:req.userDetails.userInformation.userId,
+                    stats:req.query.stats
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1140]: if stats is not there it will give error @borkarsaish65

File: generics/middleware/authenticator.js [hunk 10]
```diff
@@ -61,7 +61,7 @@ module.exports = async function (req, res, next, token = "") {
     return;
   }
 
-  let internalAccessApiPaths = ["/templates/bulkCreate"];
+  let internalAccessApiPaths = ["/templates/bulkCreate",'/userProjects/listUserProjects'];
```
Comment (ankit1shahu) [inline generics/middleware/authenticator.js:64]: you have added different endpoint here @borkarsaish65

File: module/userProjects/helper.js [hunk 11]
```diff
@@ -3273,6 +3273,63 @@ module.exports = class UserProjectsHelper {
...
+     * @param {Object} args - Query object for fetching project details.
+     * @returns {Array} - A promise that resolves to an array of project details matching the query.
+     */
+    static userProjectOverview(args,fields='all',stats) {
+          return new Promise(async (resolve, reject) => {
+            try{
+    
+                const projectIdQuery = {
+                    ...args
+                  };
```
Comment (ankit1shahu) [inline module/userProjects/helper.js:3289]: please use formatter to formate this code @borkarsaish65

File: module/userProjects/helper.js [hunk 12]
```diff
@@ -3273,6 +3273,63 @@ module.exports = class UserProjectsHelper {
...
+
+                let projects = await projectQueries.projectDocument(projectIdQuery,fieldarray);
+                
+                resolve(projects)
+            }catch(error){
+                reject(error)
+            }
+          });
+    }
+    static listUserProjects({userId,stats}){
```
Comment (ankit1shahu) [inline module/userProjects/helper.js:3314]: This function is called when? @borkarsaish65

Comment (ankit1shahu) [reply module/userProjects/helper.js:3314]: and why is this needed ? @borkarsaish65

File: controllers/v1/userProjects.js [hunk 13]
```diff
@@ -1125,5 +1126,33 @@ module.exports = class UserProjects extends Abstract {
...
+    async userProjects(req) {
+        return new Promise(async (resolve, reject) => {
+            try {
+
+                if (req.query.stats) {
+                    //convert req.query.stats string to Boolean value
+                    req.query.stats = UTILS.convertStringToBoolean(req.query.stats);
+                }
+
+                let listOfCreatedProjects = await userProjectsHelper.listUserProjects({
```
Comment (ankit1shahu) [inline controllers/v1/userProjects.js:1138]: from here you can call userprojectoverview function @borkarsaish65

**Review by ankit1shahu — CHANGES_REQUESTED** [review]

File: module/userProjects/helper.js (hunk 12 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js:3315]: it calls the userOverviewProject function by querying the required data

File: controllers/v1/userProjects.js (hunk 13 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js]: done that now

File: module/userProjects/helper.js (hunk 12 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js]: removed it now

**Review by ankit1shahu — APPROVED** [review]

**Pull request merged**